src/metabase/db/liquibase.clj

```diff
@@ -169,9 +174,10 @@
   skip creating and releasing migration locks, which is both slightly dangerous and a waste of time when we won't be
   using them.
 
-  (I'm not 100% sure whether `Liquibase.update()` still acquires locks if the database is already up-to-date)"
-  [^Liquibase liquibase]
-  (.listUnrunChangeSets liquibase nil (LabelExpression.)))
+  IMPORTANT: this function takes `data-source` but not `liquibase` because `.listUnrunChangeSets` is buggy. See #38257."
+  [^DataSource data-source]
+  (with-liquibase [liquibase (.getConnection data-source)]
+    (.listUnrunChangeSets liquibase nil (LabelExpression.))))
 
 (defn- migration-lock-exists?
```

Comment:
yea, from what I see we use `!` strictly for functions that have side-effect, we'll enforce this even further in tests see https://github.com/metabase/metabase/pull/37126

I'm open to idea for convention of this kind tho.

Reply:
A pure function will always return a value, so a function that can throw is impure. Whether you say that this is a side effect is I guess debatable, but it is a "dangerous" function as it can fail.

Since throwing and catching exceptions is [very expensive on the JVM](https://wiki.c2.com/?CheckDontCatch), typically considered bad form for control flow, I like to indicate that kind of effect in the name somehow. Perhaps the bang is a poor choice, because it can mean a lot of other things, most commonly writing to the database, especially if it's not within a transaction. 

Since this codebase frequently uses exceptions for control flow and consists largely of impure code, perhaps its premature to introduce a new convention.

I still think we should incorporate `wait` somehow in the name to mitigate surprise. How about `wait-if-migration-lock-held`? A bit clumsy still, but `wait-for-migration-lock` would imply the lock is also taken.

Reply:
which function we're talking about again? I don't think `migration-lock-exists?` blocks right? 

or are you talking about `wait-for-migration-lock-to-be-cleared` which already prefix with `wait`?

> https://wiki.c2.com/?CheckDontCatch

good to know!

Reply:
> which function we're talking about again?

Wow, such a long thread for me not noticing a fold between the name of one function and the implementation of another one 😆